sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -122,4 +122,193 @@ trait ConstraintHelper {
     case _: NullIntolerant => expr.children.flatMap(scanNullIntolerantAttribute)
     case _ => Seq.empty[Attribute]
   }
+
+  def normalizeAndReduceWithConstraints(expression: Expression): Expression =
```

Comment:
I think we could extend constraint propagation (`normalizeAndReduceWithConstraints()`) to Joins in a future PR. I believe it could work a bit similar to `inferAdditionalConstraints()` does in `constraints` and also in `InferFiltersFromConstraints` rule.

If you think this is too early to have the logic in `QueryPlanConstraints` then I can move it into `FilterReduction`.

Reply:
I've moved all new filter reduction related logic to `FilterReduction`.